Pulling from OpenMRS

kaapstorm opened the pull request:
This change pulls from OpenMRS using its Reporting API.

At the moment this imports from a single OpenMRS instance per domain. To do next: Support more than one, and associate each with a location.

Labelling "Hold off" pending a bit more testing locally.

@dannyroberts buddy @snopoke

kaapstorm committed:
Task for pulling from OpenMRS

kaapstorm committed:
View for saving config

kaapstorm committed:
Update settings.py

snopoke commented on corehq/motech/openmrs/dbaccessors.py:
```diff
@@ -0,0 +1,13 @@
+        include_docs=True,
+        reduce=False,
+    ).all()
+    return [OpenmrsImporter.wrap(result['doc']) for result in results]
```
If you use `OpenmrsImporter.view(...)` (without the `get_db()`) then the wrapping is done for you

snopoke commented on corehq/motech/openmrs/forms.py:
```diff
@@ -13,3 +22,69 @@ def __init__(self, *args, **kwargs):
+                # Don't save it if it hasn't been changed. Use simple symmetric encryption. We don't need it to be
+                # strong, considering we'd have to store the algorithm and the key together anyway; it just
+                # shouldn't be plaintext.
+                importer.password = b64encode(bz2.compress(self.cleaned_data['password']))
```
It should be possible to use the Django secret key or else the HQ RSA public key to make this more secure. 
@dannyroberts any thoughts on this?

kaapstorm replied:
(For a little context, I used the same approach to store DHIS2 and OpenClinica passwords, but repeaters store passwords in plaintext (including the OpenMRS repeater). I guess if unfriendly people are reading our database we have more things to worry about than the OpenMRS password, but I do like the idea of using the Django secret key -- it seems a little safer, and still simple.)

snopoke commented on corehq/motech/openmrs/views.py:
```diff
@@ -112,3 +117,38 @@ def openmrs_test_fire(request, domain, repeater_id, record_id):
+    def dispatch(self, request, *args, **kwargs):
+        if not toggles.OPENMRS_INTEGRATION.enabled(request.domain):
+            raise Http404()
+        return super(OpenmrsImporterView, self).dispatch(request, *args, **kwargs)
```
You can replace this with class level decorators:
```
@method_decorator(domain_admin_required, name='dispatch')
@method_decorator(toggles.OPENMRS_INTEGRATION.required_decorator(), name='dispatch')
class OpenmrsImporterView(BaseAdminProjectSettingsView):
    ....
```

snopoke commented on corehq/motech/openmrs/views.py:
```diff
@@ -112,3 +117,38 @@ def openmrs_test_fire(request, domain, repeater_id, record_id):
+    page_title = ugettext_lazy("OpenMRS Importers")
+    template_name = 'openmrs/importers.html'
+
+    @method_decorator(domain_admin_required)
```
this shouldn't be needed

dannyroberts commented on corehq/motech/openmrs/openmrs_config.py:
```diff
@@ -41,6 +49,21 @@ def get_value(self, case_trigger_info):
         return self.value
 
 
+class CasePropertyConcept(CaseProperty):
```
Is this class a separate feature, or is it added for this import-from-openmrs feature? (Fine if this PR contains a couple different things, just trying to understand.)

kaapstorm replied:
You're right @dannyroberts this is separate. I think I'll move it to a separate PR, not just to make this less confusing, but also because it would be useful to have it deployed sooner.

dannyroberts replied:
Ah, okay cool. Makes total sense. Thanks!

dannyroberts reviewed (commented):
Approach looks good to me!

kaapstorm committed:
Support locations, report params

kaapstorm committed:
Move column config to Importer model

kaapstorm committed:
Support type casting

kaapstorm committed:
Use Django secret key and AES to encrypt password

dannyroberts commented:
Saw some work on this a few days ago. Still more work to go or is it ready for another review and merge?

kaapstorm commented:
Thanks for the ping @dannyroberts. I fixed some bugs, and rebased to clean up the commits a bit. This has been tested on staging for a few days, and I think what's here is good to go.

There is one addition in the near future: An "Import Now" button on the config page that kicks off the task to import patients. I was going to include it in this PR, but I'll PR it separately.

kaapstorm committed:
Ensure AES key is not too long

kaapstorm commented:
:white_check_mark: Fixed bug with len(AES_key) > 32
